### PR TITLE
feat(nextgen): SmartEat

### PR DESCRIPTION
**Pull request by Ell1ott**
### Silent offhand
Essentially adds a extra offhand for food, so you can use your main offhand for things like totem of undying. As well works on 1.8 servers that don't support a offhand
### AutoEat
Automatically eats when the hunger gets too low

## todo:

- [x] Silent offhand
- [x] Update Comments in module
- [x] Add auto eat feature

### CODE REVIEW

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 1]
```diff
@@ -51,47 +51,50 @@ object ModuleSmartEat : Module("SmartEat", Category.PLAYER) {
 
 
 
+    private val swapBackDelay by int("SwapBackDelay", 5, 1..20)
+
+    private val preferGappleHealth by float("PreferGappleHealth", 9f, 0f..20f)
+    private val preferNotchAppleHealth by float("PreferNotchAppleHealth", 2f, 0f..20f)
+    private val preferHealthPotHealth by float("PreferHealthPotHealth", 12f, 0f..20f)
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: Those names should be a bit more self-explainatory

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                return@findBestItem 0
+
+            if(prefersGapples && item == Items.GOLDEN_APPLE)
+                return@findBestItem 100 - preferGappleHealth.toInt()
+
+            if (prefersNotchApple && item == Items.ENCHANTED_GOLDEN_APPLE)
+                return@findBestItem 100 - preferNotchAppleHealth.toInt()
+
+            foodComp.hunger
+        }
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: 1. This should not be a property, rather a function (`findFoodSlotsInHotbar` or something)
2. Using ints to refer to item slots is deprecated since it causes uneccessary bugs. Use `ItemSlot`s instead: https://github.com/CCBlueX/LiquidBounce/blob/63d9d141e73d25e2bd18ccdf81f93e7414753a39/src/main/kotlin/net/ccbluex/liquidbounce/utils/item/InventoryUtils.kt#L48 Maybe add a version of `ALL_SLOTS_IN_INVENTORY` for hotbar like `HOTBAR_SLOTS_IN_INVENTORY`.
3. The comparision part should be it's own function like `getFoodItemScore(ItemStack)`, so the code become more intuitive

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                    val scaledHeight = dc.scaledWindowHeight
+                    val i: Int = scaledWidth / 2
+                    val x = i - 91 - 26 - offset
+                    val y = scaledHeight - 16 - 3
+                    dc.drawItemInSlot(mc.textRenderer, currentFood.second, x, y)
+                    dc.drawItem(currentFood.second, x, y)
+                    dc.drawGuiTexture(
+                        HOTBAR_OFFHAND_LEFT_TEXTURE, i - 91 - 29 - offset,
+                        scaledHeight - 23, 29, 24
+                    )
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: Extract those magic numbers (like 26, 91 or 3) to constants (like `XYZ_PADDING`, `ZX_MARGIN`) so a reader understands what those numbers represent

File: src/main/kotlin/net/ccbluex/liquidbounce/utils/client/SilentHotbar.kt [hunk 4]
```diff
@@ -48,6 +48,8 @@ object SilentHotbar : Listenable {
         }
     }
 
+    fun isSlotModified(requester: Any?) = hotbarState?.requester == requester
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/utils/client/SilentHotbar.kt:53]: Missing documentation

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt (hunk 2 above)
Comment (Ell1ott) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: > This should not be a property, rather a function (findFoodSlotsInHotbar or something)

No problem, but i actually got inspired from this code: https://github.com/CCBlueX/LiquidBounce/blob/61099831d1a31961f66a241d0ae51d4ba7b67f6b/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/nofall/modes/MLG.kt#L57-L62

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt (hunk 3 above)
Comment (Ell1ott) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: I see where you are coming from, but I don't know if that really is the optimal solution. I tried to simply replicate Minecraft code as exactly as possible, meaning that the only reason the code should be changed is if minecraft would change their code. At this point, the easiest way of updating Liquidbounce's code would be to check in Minecraft's code again and see what has changed. Here it would just be confusing Liquidbounce's version of the code would use variables whilst Minecraft's wouldn't.

**Comment by 1zun4** [comment]
Any updates on this, what is missing?

**Comment by Ell1ott** [comment]
> Any updates on this, what is missing?

I was waiting for a reply from @superblaubeere27. Other than merging, everything should be ready.

**Comment by Ell1ott** [comment]
One thing that's still not perfect is that Sometimes it won't trigger silentoffhand, like when right clicking with an empty hand.

**Comment by superblaubeere27** [comment]
> One thing that's still not perfect is that Sometimes it won't trigger silentoffhand, like when right clicking with an empty hand.

If you have trouble testing, you could try testing our test framework:
- https://github.com/CCBlueX/LiquidBounce/blob/nextgen/src/test/kotlin/net/ccbluex/liquidbounce/integrationtest/TestInvCleaner.kt
- https://github.com/CCBlueX/LiquidBounce/blob/nextgen/src/test/kotlin/net/ccbluex/liquidbounce/integrationtest/TestScaffold.kt
- https://github.com/CCBlueX/LiquidBounce/blob/nextgen/src/test/kotlin/net/ccbluex/liquidbounce/integrationtest/TenaccProvider.kt

**Comment by EclipsesDev** [comment]
any progress/update here?

**Comment by Ell1ott** [comment]
My apologies for the late response. While it's not flawless yet, I haven't encountered any significant issues with it. Therefore, I'm opening it for merge now. I may improve it at a later time, but i sadly just don't currently have the time for it.

**Comment by EclipsesDev** [comment]
@1zun4

**Comment by 1zun4** [comment]
Yes, sorry! I will check it out soon.

**Comment by Ell1ott** [comment]
> Yes, sorry! I will check it out soon.

Thanks!

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+import kotlin.math.max
+
+/**
+ * SmartEat module
+ *
+ * Makes it easier to eat
+ */
+
+object ModuleSmartEat : Module("SmartEat", Category.PLAYER) {
+    private val HOTBAR_OFFHAND_LEFT_TEXTURE = Identifier("hud/hotbar_offhand_left")
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt:53]: This texture does not seem to be included in the pull request. Additionally, it should be in the `"liquidbounce"`-namespace (`Identifier("liquidbounce", "hud/hutbar_offhand_left")` or something like that)

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+    private val swapBackDelay by int("SwapBackDelay", 5, 1..20)
+
+    private val preferGappleHealth by float("MaxPreferGappleHealth", 9f, 0f..20f)
+    private val preferNotchAppleHealth by float("MaxPreferNotchAppleHealth", 2f, 0f..20f)
+    private val preferHealthPotHealth by float("MaxPreferHealthPotHealth", 12f, 0f..20f)
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: I think those names don't convey the function of the settings while being pretty verbose. I couldn't derive their full function from the code either.

Something like `PreferGappleHeathThreshold` with a fitting description (see translation files) would fit better

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+            return 0
+
+        if(prefersGapples && item == Items.GOLDEN_APPLE)
+            return 100 - preferGappleHealth.toInt()
+
+        if (prefersNotchApple && item == Items.ENCHANTED_GOLDEN_APPLE)
+            return 100 - preferNotchAppleHealth.toInt()
+
+        return foodComp.hunger
+    }
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: I see a few problems with the function:
1. `detekt` complains about the ReturnCount. In this case the complaint is appropriate. A single "lifted" return with a `when` is the way to go since it makes the code a lot more readable
2. After thinking about the code for some time I found out why you used the threshold values (i.e. `preferGappleHealth`) in this function. I think you want to prioritize the least health value. If you do such hacks comments would be great explaining why the code looks like it does. Otherwise, somebody might change it and while introducing new/old problems.
3. This entire function can be replaced by a comparator chain which would allow for more complex tie-breaking in the future.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt (hunk 3 above)
Comment (superblaubeere27) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: Yes! In this case you are right.

In this case add a comment to the code explaining what the code tries to replicate should be enough.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+            if(
+                !player.canConsume(false)
+                && currentFood.second.item.foodComponent?.isAlwaysEdible() == false
+                ) {
+                return@handler
+            }
+
+
+            SilentHotbar.selectSlotSilently(this@SilentOffhand, currentFood.first, max(swapBackDelay, 5))
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: Tip: `swapBackDelay.coerceAtMost(5)`.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+            if(player.hungerManager.foodLevel < minHunger) {
+
+                waitUntil {
+                    eat()
+                    player.hungerManager.foodLevel > minHunger
+                }
+                KeyBinding.setKeyPressed(mc.options.useKey.boundKey, false)
+
+            }
```
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: Formatting

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt (hunk 7 above)
Comment (superblaubeere27) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt]: I changed the function for you btw

**Comment by 1zun4** [comment]
@superblaubeere27 Merge this when you think it is ready.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt (hunk 5 above)
Comment (superblaubeere27) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/combat/ModuleSmartEat.kt:53]: This needs some attention from the Frontend :o @SenkJu

**Comment by superblaubeere27** [comment]
Thanks for your contribution!
